### ISSUE TITLE
Chrome Apollo Client Extension does not see schema for local state.

### ISSUE THREAD

**Issue by y-murashka** [apollographql/apollo-client-devtools#239]
Hi there. Basically i've started doing some pet project and decided not to create a server itself but use local state management for now.

I'm doing the following thing:

```javascript
const cache = new InMemoryCache();

const client = new ApolloClient({
  cache,
  resolvers,
  typeDefs
});

cache.writeData({
  data: {
    appState: {
      isCollapsed: true,
      __typename: 'AppState'
    }
  },
});
const applicationContainer = document.getElementById('application-root');
const application = (
  <ApolloProvider client={client}>
    <Pages></Pages>>
    </ApolloProvider>
);
ReactDOM.render(application, applicationContainer);
```
Here is schema:
```javascript
export const typeDefs = gql`
   type Query {
    appState: AppState
  }

   type AppState {
        isCollapsed: Boolean
    }
`;
```
And here is resolvers 
```javascript
export const GET_CART_ITEMS = gql`
  query GetAppState {
    appState @client
  }
`;

export const resolvers = {
  Query:{
    appState: ( _, { cache }) => {
      const { cartItems } = cache.readQuery({ query: GET_CART_ITEMS });
      return cartItems;
    },
  }
};
```
When i run Apollo Dev tools in chrome i expect to see schema but it's empty. Query works as expected.

![image](https://user-images.githubusercontent.com/6003658/70109757-ea0e7c00-161a-11ea-80fe-363e72752b75.png)

**Comment by dvakatsiienko** [comment]
singing up upon this. apollo local state visualization is a must have in devtools.

**Comment by ysz** [comment]
+1

**Comment by t-lock** [comment]
+2

**Comment by dvakatsiienko** [comment]
without high quality local state management monitoring (and control) this feature  doubtfully can be used as mature one in Apollo Client...although containing entire state in a cache sounds good (without Redux for instance), but again a reliable instrument to do so is required

**Comment by MichaelDM** [comment]
Same here. I am very enthusiastically contemplating using Apollo Client for all data (remote + local) but having the introspection break once I add a local schema is making me seriously reconsider this approach, as it makes dev (especially across a team) more labor intensive. Are you guys contemplating a fix for this?

**Comment by dvakatsiienko** [comment]
I was testing Apollo Local State management for a REALLY simple usecase:

+ isAuthenticated flag;
+ Application Authorized Endpoints list;

And after tinkering with it for ~3 weeks or so I've decided to delete all local state from Apollo and move it to Redux. And I am happy now. As for now Apollo client local state it not comfortable to work with due to its verbosity and lack of possibility for simple acces and visualization. Although the idea is very cool.

**Comment by ElForastero** [comment]
Does anyone have any ideas on how to fix this?

**Comment by MichaelDM** [comment]
Hey guys, 
Update on my end. 
I actually figured out it wasn't an error on apollo's side, but on mine. 
I had a mistake in my typeDefs file. 
Even though the typeDefs had more complexity than the example of local Management in apolloGraphql example (https://www.apollographql.com/docs/tutorial/local-state/), it worked like a charm. 
@ElForastero, might you post (and/or double-check) your typeDefs? Perhaps, as I did, you have an issue there?

**Comment by ElForastero** [comment]
@MichaelDM Which error did you have?

I have defined typeDefs as a regular *.graphql file with client schema. So it is validated by de IDE and codegen. I can't see any errors here. I'm starting to suspect, that this is because I don't have any remote endpoints right now. I have only local state and therefore all my schema constists of local queries/mutations and resolvers.

<details>
<summary>client.schema.graphql</summary>

```graphql
type Source {
  id: ID!
  url: String!
  link: String!
  title: String!
  description: String
  image: String
  color: String
}

input SourceInput {
  url: String!
  link: String!
  title: String!
  description: String
  image: String
  color: String
}

type Feed {
  id: ID!
  # Unique identifier of the feed
  # <guid> in the RSS, <id> in the Atom
  guid: String!
  # External key
  source_id: ID!
  link: String!
  title: String!
  description: String
  content: String
  pubDate: String
  author: String
  image: String
  categories: [String]
  seen: Boolean!
}

input FeedInput {
  guid: String!
  source_id: ID!
  link: String!
  title: String!
  seen: Boolean
  description: String
  content: String
  pubDate: String
  author: String
  image: String
  categories: [String]
}

type Query {
  sources: [Source]!
  feed(sourceID: ID): [Feed]!
}

type Mutation {
  addSource(input: SourceInput!): Source!
  markFeedAsSeen(id: ID!): Feed!
}
```
</details>

<details>
<summary>client.ts</summary>

```ts
import { ApolloClient, InMemoryCache } from '@apollo/client';
import { resolvers as sources } from './resolvers/sources';
import { resolvers as feed } from './resolvers/feed';
import typeDefs from './schema.client.graphql';

const cache = new InMemoryCache();

export const client = new ApolloClient({
  cache,
  typeDefs,
  resolvers: {
    Query: {
      ...sources.Query,
      ...feed.Query,
    },
    Mutation: {
      ...sources.Mutation,
      ...feed.Mutation,
    },
  },
});
```
</details>

**Comment by bennypowers** [comment]
Same problem, but wasn't solved by adding a backend.

https://spectrum.chat/apollo/apollo-link-state/client-has-typedefs-but-no-docs-in-graphiql~8a80e439-3257-4fce-a666-a1e75c3cec63

**Comment by MichaelDM** [comment]
@ElForastero here is an idea.
I have at minimum the extend keyword:
`extend type Query { ... }`
` extend type Mutation{ ...}`
Would have to look over the docs again, but probably need to apply it well to your other Types.

**Comment by ElForastero** [comment]
@MichaelDM I've tried the `extend` keyword and nothing was changed.

**Comment by Remownz** [comment]
same here: minimum typeDefs with or without "extend" keyword but no schema documentation inside devtools

**Comment by sandorTuranszky** [comment]
Does not work for me either. Any comment on this from the Apollo team?

**Comment by MichaelDM** [comment]
What I think looking back was most frustrating for me, was the lack of error messages to help me debug the issues I had. 

Unfortunately, @ElForastero , @Remownz , there are no errors that are apparent to my eyes when I read your code snippets. 

This said, how I would go about solving this, is going from an implementation that works (even if it's the tutorial example), modifying it bit by bit to get to the complexity that you have today, and see where it breaks in the process. If you find the solution, please post it here because I'm sure others will find it useful! 

Off the top of my head, when I did it on my end, I realized I had typos in my schema, like `query` lowercase, or writing a wrong type (Number) instead of Int...

**Comment by bennypowers** [comment]
I've gotten to the point where the VS Code extension "Apollo GraphQL" properly loads and executes my client-side schema. I did this by running a code-generation script in a file watcher:

```js
import { mergeTypeDefs } from '@graphql-tools/merge';
import { loadFilesSync } from '@graphql-tools/load-files';
import { writeFileSync } from 'fs';
import { print } from 'graphql';

const types =
  loadFilesSync('./src/**/*.client.schema.graphql');

const merged =
  print(mergeTypeDefs(types)).replace(`schema {
  query: Query
  mutation: Mutation
}`, '');

writeFileSync('./src/apollo/schema/generated.schema.graphql', merged);
```

So now I get autocomplete in my component graphql files, great!

However, I still don't get GraphiQL working with my client schema, only the server schema via introspection query.

**Comment by Natas0007** [comment]
Any updates on this issue from Apollo?  I'm running the latest Apollo Client v3.1.4 and no typeDefs defined on the client side are shown in Dev Tools Schema tab (DevTools v2.3.1).  I've tried with and without the extend keyword in front of type.  Server side typeDefs are shown fine.  Even when completely disconnecting the server in the config (for testing), no client typeDefs (schema) are shown.

**Comment by tobiasriemenschneider** [comment]
+1

**Comment by reifnotreef** [comment]
+1

Just had this start happening to me yesterday.

**Comment by dstockdale** [comment]
+1

**Comment by skonves** [comment]
A workaround to manually revert back to v2.3.1:

1. Clone this repo
1. Checkout the previous release tag: `git checkout v2.3.1`
1. Install deps: `npm ci`
1. Build the extension: `npm run build`
1. Install the build output as an extension in Chrome
    1. ensure the "real" extension is uninstalled
    1. Nav to `chrome://extensions/`
    1. Switch on "Developer mode" in the top right corner
    1. Click the "Load unpacked" button in the top left corner
    1. Select the output file (it will be in the `.../shells/webextension` folder wherever you have cloned the repo)
1. Thank the dev team here for fantastic automation and a well-managed project that makes this possible 🎉

**Comment by vigie** [comment]
This fix works well for me: configure the client to use this "IntrospectionLink". You can get [GraphQL Code Generator](https://www.graphql-code-generator.com/docs/plugins/introspection) to generate the introspectionResult for you from your local schema definition.

Tested with devtools v3.0.2, works well:

```typescript
import { ApolloLink, NextLink, Operation } from '@apollo/client/core';
import { ExecutionResult } from 'graphql';
import Observable from 'zen-observable';

import introspectionResult from './schema-introspection.json';

/**
 * Serves introspection operations. For example, the Apollo Client
 * Chrome Devtool issues an introspection operation when it opens
 * in order to display the schema.
 */
export class IntrospectionLink extends ApolloLink {
    request(operation: Operation, forward?: NextLink) {
        switch (operation.operationName.toLowerCase()) {
            case 'introspectionquery':
                return new Observable<ExecutionResult>((subscriber) => {
                    subscriber.next({
                        data: introspectionResult,
                    });
                    subscriber.complete();
                });
        }

        if (forward) {
            return forward(operation);
        }

        throw new Error(`Unable to handle operation ${operation.operationName}`);
    }
}
```

**Comment by reifnotreef** [comment]
I think the extension has recently updated, the UI is quite different. I can now see all current page queries/mutation in the tabs as intended, as well as my schema etc.  

I believe I couldn't see them before because we were on a beta version of Apollo that was not supported by the extension yet.

**Comment by hwillson** [comment]
@reifnotreef yes, we just launched 3.0 which includes the new UI you've noticed, but is also a full re-write behind the scenes to put a better foundation in place for future growth. Glad to hear it's working - thanks!

**Comment by vigie** [comment]
Yes, really enjoying the update, thanks team! But note it does _not_ fix the subject of this issue - showing local only schemas. My above workaround will get you there in the meantime.

**Comment by hwillson** [comment]
For sure, thanks @vigie - we'll dig into this shortly.

**Comment by adityasingh773** [comment]
I am trying to follow the local Client Side Schema as per this tutorial: https://www.apollographql.com/docs/react/local-state/client-side-schema/
Here is my schema 
`const typeDefs = gql`
  extend type Query {
    isLoggedIn: Boolean!
    cartItems: [Launch]!
  }

  extend type Launch {
    isInCart: Boolean!
  }

  extend type Mutation {
    addOrRemoveFromCart(id: ID!): [Launch]
  }
`;`
As you can see this is a very simple use case, a bare example. But I don't see anything being reflected on my Apollo Dev Tools Docs.
What's wrong?

**Comment by rektide** [comment]
I mistook this for a thread on seeing reactive variables. There is a feature request for that at: https://github.com/apollographql/apollo-client-devtools/discussions/391

**Comment by cortopy** [comment]
I'm using 3.0.5 in chromium and it's broken for me. I'm only following the tutorial for local state

This is my local schema

```graphql
extend type RootQuery {
    isLoggedIn: Boolean!
  }
```

If I use the @client directive, it tells me that's not right

![image](https://user-images.githubusercontent.com/8725825/125277489-642fd800-e309-11eb-8a0c-20b73bef4aeb.png)

If I don't have the directive and select "Load from cache" I still get the data though
![image](https://user-images.githubusercontent.com/8725825/125277580-84f82d80-e309-11eb-8fff-39822122c16e.png)

So it seems like an issue with devTools, not with apollo client. The cache tab also shows that the cache is empty, even though it's not

**Comment by spigelli** [comment]
Also still seeing this
